drivers/bluetooth/hci/spi.c

```diff
@@ -291,7 +291,7 @@ static int bt_spi_send_aci_config_data_controller_mode(void)
 	/* Force BlueNRG-MS roles to Link Layer only mode */
 	param->value = BLUENRG_ACI_LL_MODE;
 
-	bt_hci_cmd_send(BLUENRG_ACI_WRITE_CONFIG_DATA, buf);
+	hci_cmd_send(BLUENRG_ACI_WRITE_CONFIG_DATA, buf);
```

Comment:
I'm not following how this is supposed to work (at the same time I don't see CI failures so maybe I'm missing something): you're putting the declarations of `hci_cmd_create()` and `hci_cmd_send()` into the host-internal header file `hci_core.h` but I don't see the drivers including it anywhere? To be fair, they probably shouldn't, since that header file is supposed to only be used by code in `subsys/bluetooth/host/*`.

Originally, all these HCI APIs were strictly internal to the host, i.e. they were in `hci_core.h`. At some point we acknowledged the need to use them from drivers for various purposes (which is what you see here) and because of that we needed to open them up through `include/bluetooth/hci.h`. Unless I'm missing something this patch is reverting that? Also, since these are primarily for testing/hacking/debugging in the app, I'd question a bit the need to make "safe" versions of them. I guess it'd be good to have some more discussion around this.

If your intention is to get something still fixed for 2.1 (I have a feeling this might be the case since @carlescufi pinged me on this PR) I'd suggest to stick to strictly "critical" items, i.e. things which are known to be broken or confusing in existing code. Those will be easier to agree on and get merged.

Reply:
I am puzzled by this too. I built a sample for `96b_carbon` which uses `spi.c` and GCC doesn't emit a warning.

Reply:
Ah wait, that's only for BlueNRG. Now I built another sample and I indeed get the error:

```
$ west build -b nucleo_wb55rg samples/bluetooth/beacon/
-- west build: build configuration:
       source directory: /home/carles/src/zephyr/zephyr/samples/bluetooth/beacon
       build directory: /home/carles/src/zephyr/zephyr/build (created)
       BOARD: nucleo_wb55rg (origin: command line)
-- west build: generating a build system
Zephyr version: 2.1.0
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.7.5", minimum required is "3.4")
-- Selected BOARD nucleo_wb55rg
-- Found west: /home/carles/.local/bin/west (found suitable version "0.6.3", minimum required is "0.6.0")
-- Loading /home/carles/src/zephyr/zephyr/boards/arm/nucleo_wb55rg/nucleo_wb55rg.dts as base
Devicetree configuration written to /home/carles/src/zephyr/zephyr/build/zephyr/include/generated/generated_dts_board.conf
Parsing Kconfig tree in /home/carles/src/zephyr/zephyr/Kconfig
Loaded configuration '/home/carles/src/zephyr/zephyr/boards/arm/nucleo_wb55rg/nucleo_wb55rg_defconfig'
Merged configuration '/home/carles/src/zephyr/zephyr/samples/bluetooth/beacon/prj.conf'
Configuration saved to '/home/carles/src/zephyr/zephyr/build/zephyr/.config'
-- The C compiler identification is GNU 8.3.1
-- The CXX compiler identification is GNU 8.3.1
-- The ASM compiler identification is GNU
-- Found assembler: /home/carles/bin/gcc-arm-none-eabi-8-2019-q3-update/bin/arm-none-eabi-gcc
-- Cache files will be written to: /home/carles/.cache/zephyr
-- Configuring done
-- Generating done
-- Build files have been written to: /home/carles/src/zephyr/zephyr/build
-- west build: building application
[1/154] Preparing syscall dependency handling

[58/154] Building C object zephyr/CMakeFiles/zephyr.dir/drivers/bluetooth/hci/ipm_stm32wb.c.obj
/home/carles/src/zephyr/zephyr/drivers/bluetooth/hci/ipm_stm32wb.c: In function 'bt_ipm_set_addr':
/home/carles/src/zephyr/zephyr/drivers/bluetooth/hci/ipm_stm32wb.c:415:8: warning: implicit declaration of function 'hci_cmd_create'; did you mean 'bt_hci_cmd_create'? [-Wimplicit-function-declaration]
  buf = hci_cmd_create(ACI_HAL_WRITE_CONFIG_DATA, sizeof(*param));
        ^~~~~~~~~~~~~~
        bt_hci_cmd_create
/home/carles/src/zephyr/zephyr/drivers/bluetooth/hci/ipm_stm32wb.c:415:6: warning: assignment to 'struct net_buf *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
  buf = hci_cmd_create(ACI_HAL_WRITE_CONFIG_DATA, sizeof(*param));
      ^
/home/carles/src/zephyr/zephyr/drivers/bluetooth/hci/ipm_stm32wb.c:431:8: warning: implicit declaration of function 'hci_cmd_send_sync'; did you mean 'bt_hci_cmd_send_sync'? [-Wimplicit-function-declaration]
  err = hci_cmd_send_sync(ACI_HAL_WRITE_CONFIG_DATA, buf, &rsp);
        ^~~~~~~~~~~~~~~~~
        bt_hci_cmd_send_sync
/home/carles/src/zephyr/zephyr/drivers/bluetooth/hci/ipm_stm32wb.c: In function 'bt_ipm_ble_init':
/home/carles/src/zephyr/zephyr/drivers/bluetooth/hci/ipm_stm32wb.c:457:6: warning: assignment to 'struct net_buf *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
  buf = hci_cmd_create(ACI_WRITE_SET_TX_POWER_LEVEL, 3);
      ^
[149/154] Linking C executable zephyr/zephyr_prebuilt.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       38624 B       812 KB      4.65%
            SRAM:       11684 B        96 KB     11.89%
           SRAM1:         424 B        10 KB      4.14%
           SRAM2:        2440 B        20 KB     11.91%
        IDT_LIST:         152 B         2 KB      7.42%
[154/154] Linking C executable zephyr/zephyr.elf
```

Reply:
@jhedberg @carlescufi  Ok. I made the mistake of thinking that the drivers used the internal header file.

> Also, since these are primarily for testing/hacking/debugging in the app, I'd question a bit the need to make "safe" versions of them. I guess it'd be good to have some more discussion around this.

We are using this to send Vendor Specific HCI commands atleast, maybe it is not the right way? But for that reason I thought it appropriate to create a public version that checked the ready flag.

The HCI drivers cannot use this version however. Since they are called during hci driver open() the flag is not set.

I think we have these options:
1. Revert the commit
2. add hci_driver_cmd_ to hci_driver.h that gives access to the internal functions
3. Add different flag and use bt_hci_cmd_ in the drivers

Thoughs?

Reply:
There are two technical pre-requisites for sending commands:

1. The host must have its structures & threads initialized
2. The HCI driver must have its HCI transport functional on some basic level

For the host, when drv->open() returns it is safe to assume that the transport has been set up, however as long as we're inside open() only the driver knows whether the transport is functional or not.

I'm not really a fan of adding more APIs unless we absolutely need to. To me your vendor extensions are kind of proprietary extensions of the host rather than a layer on top of it, i.e. they could potentially check the host's internal state, like the BT_DEV_READY flag, before calling the HCI APIs.

Reply:
@jhedberg Well I'd prefer not to check internal state. But I can use the bt_ready callback to set my own. So that can be handled.
Should I revert the commit then?

Reply:
@jhedberg @carlescufi Removed the HCI ready commit.